dpp/src/data_contract/state_transition/data_contract_create_transition/validation/state/validate_data_contract_create_transition_state_factory.rs

```diff
@@ -0,0 +1,47 @@
+use crate::{
+    errors::StateError, mocks, state_repository::StateRepositoryLike, validation::ValidationResult,
+};
+use anyhow::Result;
+
+pub struct ValidateDataContractCreateTransitionStateFactory<SR>
+where
+    SR: StateRepositoryLike,
+{
+    state_repository: SR,
+}
+
+pub fn fetch_documents_factory<SR>(
+    state_repository: SR,
+) -> ValidateDataContractCreateTransitionStateFactory<SR>
+where
+    SR: StateRepositoryLike,
+{
+    ValidateDataContractCreateTransitionStateFactory { state_repository }
+}
+
+impl<SR> ValidateDataContractCreateTransitionStateFactory<SR>
+where
+    SR: StateRepositoryLike,
+{
+    pub async fn validate_data_contract_create_transition_state(
+        &self,
+        state_transition: mocks::StateTransition,
+    ) -> ValidationResult {
+        let mut result = ValidationResult::default();
+
+        // Data contract shouldn't exist
+        let maybe_existing_data_contract: Result<Option<Vec<u8>>> = self
+            .state_repository
+            .fetch_data_contract(&state_transition.data_contract.id)
+            .await;
```

Comment:
I'm actually wondering how that's going to work with WASM - I don't have any experience with async and WASM, would be interesting to check it out

Reply:
Not much expirence from me either. While implementing I've learnt that the `future` is being sent out to the js event loop, so in theory, its good to have async functions instead of synchronous ones. This will require closer look for sure.